Smt/Data/Sexp.lean

```diff
@@ -23,21 +26,29 @@ def serializeMany (ss : List Sexp) : String :=
 instance : ToString Sexp :=
   ⟨serialize⟩
 
-partial def parse (s : String) : Except String (List Sexp) :=
-  let tks := tokenize #[] s.toSubstring
+partial def parse (s : String) : Except String (List Sexp) := do
+  let tks ← tokenize #[] s.toSubstring
   parseMany #[] tks.toList |>.map Prod.fst |>.map Array.toList
 where
-  tokenize (stk : Array Substring) (s : Substring) : Array Substring :=
-    if s.isEmpty then stk
-    else
-      let c := s.front
-      if c == ')' || c == '(' then
-        tokenize (stk.push <| s.take 1) (s.drop 1)
-      else if c.isWhitespace then tokenize stk (s.drop 1)
-      else
-        let tk := s.takeWhile fun c => !c.isWhitespace && c != '(' && c != ')'
-        if tk.bsize > 0 then tokenize (stk.push tk) (s.extract ⟨tk.bsize⟩ ⟨s.bsize⟩)
-        else unreachable!
+  tokenize (stk : Array Substring) (s : Substring) : Except String (Array Substring) := do
+    if s.isEmpty then
+      return stk
+    let c := s.front
+    if c == '"' then
+      let s1 := s.drop 1 |>.takeWhile (· ≠ '"')
+      if s1.stopPos = s.stopPos then
+        throw "ending \" missing"
+      let s1 := ⟨s.str, s.startPos, s.next s1.stopPos⟩
+      let s2 := ⟨s.str, s1.stopPos, s.stopPos⟩
+      return (← tokenize (stk.push s1) s2)
+    if c == ')' || c == '(' then
+      return (← tokenize (stk.push <| s.take 1) (s.drop 1))
+    if c.isWhitespace then
+      return (← tokenize stk (s.drop 1))
+    let tk := s.takeWhile fun c => !c.isWhitespace && c != '(' && c != ')'
+    if tk.bsize > 0 then
+      return (← tokenize (stk.push tk) (s.extract ⟨tk.bsize⟩ ⟨s.bsize⟩))
+    else unreachable!
```

Comment:
Any idea why I need to guard `unreachable!`? All tests fail with the error
```
PANIC at Sexp.parse.tokenize Smt.Data.Sexp:51:4: unreachable code has been reached
```
This only happens with the changes you made to `lakefile.lean` (compiling to shared libraries). Could this a bug in the compiler? Also, why did you enable compiling?

Reply:
I can't reproduce this, do you have an example of where it fails?

Reply:
hmm... I can reproduce when I run this command:
```bash
LEAN_PATH=build/lib lean --load-dynlib=build/lib/libSmt.so Temp.lean
```
`Temp.lean` contents:
```lean
import Smt
```
btw, it looks like the VS Code extension does not load dynamic libraries by default?

Reply:
If just an `import` suffices, then `unreachable!` must have been considered a closed term and extracted into an initialization block that runs on startup. Anyhow I made the parser strict [here](https://github.com/Vtec234/lean-smt/tree/solver), feel free to merge!

> Also, why did you enable precompiling?

It just makes some of tests run faster, but we could disable if it causes too much trouble.

> it looks like the VS Code extension does not load dynamic libraries by default?

It does, but it's a bit hidden. They're not passed on the command line but rather the Lean server communicates with Lake to find them.